ambry-tools/src/main/java/com.github.ambry.tools/perf/rest/PerfNioServer.java

```diff
@@ -0,0 +1,419 @@
+        readableStreamChannel = new PerfRSC(chunk, totalBlobSize);
+        addValueForHeader(inFluxArgs, RestUtils.Headers.BLOB_SIZE, Long.toString(totalBlobSize));
+        addValueForHeader(inFluxArgs, RestUtils.Headers.SERVICE_ID, "PerfNioServer");
+        addValueForHeader(inFluxArgs, RestUtils.Headers.AMBRY_CONTENT_TYPE, "application/octet-stream");
+        addValueForHeader(inFluxArgs, "x-um-perf-user-metadata", usermetadata);
+      } else {
+        readableStreamChannel = new PerfRSC(null, 0);
+      }
+      args = Collections.unmodifiableMap(inFluxArgs);
+    }
+
+    @Override
+    public RestMethod getRestMethod() {
+      return restMethod;
+    }
+
+    @Override
+    public String getPath() {
+      return PERF_PATH;
+    }
+
+    @Override
+    public String getUri() {
+      return PERF_URI;
+    }
+
+    @Override
+    public Map<String, Object> getArgs() {
+      return args;
+    }
+
+    @Override
+    public void prepare() {
+      // no op.
+    }
+
+    @Override
+    public boolean isOpen() {
+      return readableStreamChannel.isOpen();
+    }
+
+    @Override
+    public void close()
+        throws IOException {
+      readableStreamChannel.close();
+      restRequestMetricsTracker.nioMetricsTracker.markRequestCompleted();
+      restRequestMetricsTracker.recordMetrics();
+    }
+
+    @Override
+    public RestRequestMetricsTracker getMetricsTracker() {
+      return restRequestMetricsTracker;
+    }
+
+    @Override
+    public long getSize() {
+      return readableStreamChannel.getSize();
+    }
+
+    @Override
+    public Future<Long> readInto(AsyncWritableChannel asyncWritableChannel, Callback<Long> callback) {
+      return readableStreamChannel.readInto(asyncWritableChannel, callback);
+    }
+
+    @Override
+    public void setDigestAlgorithm(String digestAlgorithm)
+        throws NoSuchAlgorithmException {
+      readableStreamChannel.setDigestAlgorithm(digestAlgorithm);
+    }
+
+    @Override
+    public byte[] getDigest() {
+      return readableStreamChannel.getDigest();
+    }
+
+    /**
+     * Adds a value for a header.
+     * @param toAddTo the map to add the {@code key} with value {@code value} to.
+     * @param key the key for which {@code value} is a value.
+     * @param value a value of {@code key}
+     */
+    private void addValueForHeader(Map<String, Object> toAddTo, String key, Object value) {
+      if (value != null && value instanceof String) {
+        StringBuilder sb;
+        if (toAddTo.get(key) == null) {
+          sb = new StringBuilder(value.toString());
+          toAddTo.put(key, sb);
+        } else {
+          sb = (StringBuilder) toAddTo.get(key);
+          sb.append(MULTIPLE_HEADER_VALUE_DELIMITER).append(value);
+        }
+      } else if (value != null && toAddTo.containsKey(key)) {
+        throw new IllegalStateException("Value of key [" + key + "] is not a string and it already exists in the args");
+      } else {
+        toAddTo.put(key, value);
```

Comment:
When `value` is not a `String`, why the entry can be put into the map IF the key does not exist, but cannot otherwise?